docs/fundamentals/networking/websockets.md

```diff
@@ -0,0 +1,52 @@
+---
+title: WebSockets support in .NET
+description: Learn how to web sockets with the ClientWebSockets in .NET.
+author: greenEkatherine
+ms.author: esokolov
+ms.date: 10/27/2022
+---
+
+# WebSockets support in .NET
+
+The WebSocket Protocol enables two-way communication between a client to a remote host that has opted-in to communications from that code. The <xref:System.Net.WebSockets.ClientWebSocket?displayProperty=fullName> exposes the ability to establish WebSocket connection via an opening handshake: it is created and sent by `ConnectAsync` method.
+
+## HTTP/2 vs HTTP/1.1 WebSockets
+
+WebSockets over HTTP/1.1 uses a single TCP connection, therefore it is managed by connection-wide headers (see [RFC 6455](https://www.rfc-editor.org/rfc/rfc6455)). Here is an example of how to establish WebSocket over HTTP/1.1:
+
+```c#
+ClientWebSocket ws = new();
+ws.ConnectAsync(uri, cancellationToken);
+```
+
+A different approach must be taken with HTTP/2 due to its multiplexing nature: here, WebSockets are established per stream (see [RFC 8441](https://www.rfc-editor.org/rfc/rfc8441)). With HTTP/2 it is possible to use one connection for multiple web socket streams together with ordinary HTTP streams and extends HTTP/2's more efficient use of the network to WebSockets. There is a special overload of `ConnectAsync` which accepts `HttpMessageInvoker` to allow reusing existing pooled connections:
+
+```c#
+var handler = new SocketsHttpHandler();
+ClientWebSocket ws = new();
+ws.ConnectAsync(uri, new HttpMessageInvoker(handler), cancellationToken);
+```
+
+## How to set up HTTP version and policy
+
+By default, `ClientWebSocket` uses HTTP/1.1 to send an opening handshake and allows downgrade. It can be changed before calling `ConnectAsync`:
+
+```c#
+ClientWebSocket ws = new();
+
+ws.Options.HttpVersion = HttpVersion.Version20;
+ws.Options.HttpVersionPolicy = HttpVersionPolicy.RequestVersionOrHigher;
+
+ws.ConnectAsync(uri, new HttpMessageInvoker(handler), cancellationToken);
+```
+
+## Incompatible options
+
+`ClientWebSocket` has properties <xref:System.Net.WebSockets.ClientWebSocketOptions?displayProperty=fullName> that the user can set up before the connection is established. However, when `HttpMessageInvoker` is provided, it also has these properties. To avoid ambiguity, in that case, properties should be set on `HttpMessageInvoker`, and `ClientWebSocketOptions` should have default values. Otherwise, if `ClientWebSocketOptions` are changed, overload of `ConnectAsync` will throw an <xref:System.ArgumentException>.
+
+## Compression
```

Comment:
Can we have an example here please? It could be taken from https://devblogs.microsoft.com/dotnet/announcing-net-6-preview-5/#libraries-websocket-compression

Reply:
example on how to turn off compression for specific messages would be also very useful (so it is not only said in text here)

Reply:
> example on how to turn off compression for specific messages would be also very useful (so it is not only said in text here)

I cannot find how to do it

Reply:
Huh... Turns out we have this option available only for `WebSocket`, not `ClientWebSocket` (see `WebSocketMessageFlags` in [WebSocket.SendAsync](https://learn.microsoft.com/en-us/dotnet/api/system.net.websockets.websocket.sendasync?view=net-6.0#system-net-websockets-websocket-sendasync(system-readonlymemory((system-byte))-system-net-websockets-websocketmessagetype-system-net-websockets-websocketmessageflags-system-threading-cancellationtoken))) 🤦‍♀️🤦‍♀️🤦‍♀️ We should expose this API on `ClientWebSocket` as well. Will you be able to create an issue for that pls? Or I can do that.

Reply:
Ok, I will create an issue

Reply:
https://github.com/dotnet/runtime/issues/77783, did I miss anything? If not, I will update it with API proposal a little bit later